### ISSUE TITLE
sqlalchemy error on `make migrate-up`

### ISSUE THREAD

**Issue by uncojohnco** [alysivji/busy-beaver#81]
Related: #72 

When executing `make up`, it looks like the make migrate command is failing, specifically on `docker-compose exec app flask db upgrade.

Though the following make commands are working as expected
- make test
- make shell
- make shell-dev

I'm guessing I'm missing some sort of initial configuration? (Also, note that I'm pretty much a newbie dealing with a web stack :P)

Output from running `make up`
``` shell
johno-dev@johnodev-VirtualBox:~/PycharmProjects/busy-beaver/busy-beaver$ make up
docker-compose up -d
WARNING: The NGROK_BASE_URI variable is not set. Defaulting to a blank string.
WARNING: The SLACK_BOTUSER_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_ID variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_SECRET variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_SECRET variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_KEY variable is not set. Defaulting to a blank string.
WARNING: The DATABASE_URI variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN_SECRET variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The POSTGRES_PASSWORD variable is not set. Defaulting to a blank string.
WARNING: The POSTGRES_USER variable is not set. Defaulting to a blank string.
busy-beaver_app_1 is up-to-date
busy-beaver_db_1 is up-to-date
make migrate-up
make[1]: Entering directory '/home/johno-dev/git/busy-beaver'
docker-compose exec app flask db upgrade
WARNING: The NGROK_BASE_URI variable is not set. Defaulting to a blank string.
WARNING: The SLACK_BOTUSER_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_ID variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_SECRET variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_SECRET variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_KEY variable is not set. Defaulting to a blank string.
WARNING: The DATABASE_URI variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN_SECRET variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The POSTGRES_PASSWORD variable is not set. Defaulting to a blank string.
WARNING: The POSTGRES_USER variable is not set. Defaulting to a blank string.
[2019-03-07 00:34:14] INFO busy_beaver:29 [BusyBeaver] Configure Integrations
[2019-03-07 00:34:14] INFO busy_beaver:40 [BusyBeaver] Starting Server
Traceback (most recent call last):
  File "/usr/local/bin/flask", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 894, in main
    cli.main(args=args, prog_name=name)
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 557, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 412, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/cli.py", line 134, in upgrade
    _upgrade(directory, revision, sql, tag, x_arg)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/__init__.py", line 95, in wrapped
    f(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/__init__.py", line 280, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/usr/local/lib/python3.7/site-packages/alembic/command.py", line 254, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.7/site-packages/alembic/script/base.py", line 427, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/usr/local/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.7/site-packages/alembic/util/compat.py", line 82, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "migrations/env.py", line 95, in <module>
    run_migrations_online()
  File "migrations/env.py", line 77, in run_migrations_online
    poolclass=pool.NullPool,
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py", line 466, in engine_from_config
    return create_engine(url, **options)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py", line 425, in create_engine
    return strategy.create(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 50, in create
    u = url.make_url(name_or_url)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/url.py", line 211, in make_url
    return _parse_rfc1738_args(name_or_url)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/url.py", line 270, in _parse_rfc1738_args
    "Could not parse rfc1738 URL from string '%s'" % name)
sqlalchemy.exc.ArgumentError: Could not parse rfc1738 URL from string ''
Makefile:48: recipe for target 'migrate-up' failed
make[1]: *** [migrate-up] Error 1
make[1]: Leaving directory '/home/johno-dev/git/busy-beaver'
Makefile:35: recipe for target 'up' failed
make: *** [up] Error 2
```

**Comment by uncojohnco** [comment]
My initial entry point to investigate from the stack trace is in `.../.migrations/env.py`...
```python
...
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/__init__.py", line 280, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/usr/local/lib/python3.7/site-packages/alembic/command.py", line 254, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.7/site-packages/alembic/script/base.py", line 427, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/usr/local/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.7/site-packages/alembic/util/compat.py", line 82, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "migrations/env.py", line 95, in <module>
    run_migrations_online()
  File "migrations/env.py", line 77, in run_migrations_online
    poolclass=pool.NullPool,
...
```

**Comment by alysivji** [comment]
Was able to replicate this error. Looks like you gotta get some env variables

```console
export POSTGRES_USER=[username]
export POSTGRES_PASSWORD=[password]
export DATABASE_URI=postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:5432/busy-beaver
```

This is the bare min to get started with a development environment.

I should just throw these credentials into the docker-compose. Thoughts on having this as env variables (instructions) or just hardcoded in the docker-compose? I was leaning towards instructions, but after having both you and Cody experience errors re: database creds, I think I need to go the other route.

Thanks @uncojohnco for being my guinea pig! Apologies for the lack of clear instructions. I am moving that up the list. Will get it done today! Task queue is important, but this is even more so.

**Comment by uncojohnco** [comment]
No worries, the next step was to start populating the env vars :upside_down_face: 
 
> I should just throw these credentials into the docker-compose. Thoughts on having this as env variables (instructions) or just hardcoded in the docker-compose? I was leaning towards instructions, but after having both you and Cody experience errors re: database creds, I think I need to go the other route.

What about implementing a base `.env` file [docker-compose env file](https://docs.docker.com/compose/environment-variables/)? Then people can add credentials that they require and makes it exposed at a `root` context and tightly coupled as `config`?

I'd suggest to still have in the high-level docs what env-vars are required and what's optional. No need to go into detail, but at least it makes first-time contributors aware of the key parms that `busy-beaver` uses 

---

I look to be still running into newbie issues :stuck_out_tongue_closed_eyes: 
I currently have a `.env` in the root repo with the following:
```
POSTGRES_USER=admin
POSTGRES_PASSWORD=admin

DATABASE_URI=postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:5432/busy-beaver
```
Though when running `make up`, it seems to be tripping at `make migrate-up`, `docker-compose exec app flask db upgrade` step with the following 

```python
...
make[1]: Entering directory '/home/johno-dev/git/busy-beaver'
docker-compose exec app flask db upgrade
WARNING: The NGROK_BASE_URI variable is not set. Defaulting to a blank string.
WARNING: The SLACK_BOTUSER_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_OAUTH_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_ID variable is not set. Defaulting to a blank string.
WARNING: The GITHUB_APP_CLIENT_SECRET variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_KEY variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_CONSUMER_SECRET variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The TWITTER_ACCESS_TOKEN_SECRET variable is not set. Defaulting to a blank string.
 * Tip: There are .env files present. Do "pip install python-dotenv" to use them.
[2019-03-08 04:08:42] INFO busy_beaver:29 [BusyBeaver] Configure Integrations
[2019-03-08 04:08:42] INFO busy_beaver:40 [BusyBeaver] Starting Server
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2158, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 342, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 788, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 529, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 1284, in _do_get
    return self._create_connection()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 347, in _create_connection
    return _ConnectionRecord(self)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 474, in __init__
    self.__connect(first_connect_check=True)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/pool.py", line 671, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 106, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 412, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  role "${POSTGRES_USER}" does not exist
```
I haven't done much db migration before, but I'm checking the migrations/versions ...

running `docker-compose exec app flask db history`
```
[2019-03-08 04:12:40] INFO busy_beaver:29 [BusyBeaver] Configure Integrations
[2019-03-08 04:12:40] INFO busy_beaver:40 [BusyBeaver] Starting Server
500e1baf1bef -> c2aead9ff6d9 (head), add roles to api user
a5915c5a78eb -> 500e1baf1bef, add key value store for slack workspaces
78514b173380 -> a5915c5a78eb, remove key-value store (stop using simplekv... we will do our own thing)
312e762dfa85 -> 78514b173380, add key-vaue table store (using simplekv)
3f5a7657912d -> 312e762dfa85, create api user table
<base> -> 3f5a7657912d, make user table
```

**Comment by uncojohnco** [comment]
looks like as I was using the `.env` file the `DATABASE_URI` couldn't reference `POSTGRES_USER` var. Hence hardcoding `DATABASE_URI` in docker-compose.yml fixed this. 

I'll create a PR spike with the solve to start discussions for this implementation and see what ppls throguhs are...

**Comment by alysivji** [comment]
I'm getting a bit lost with all of the issues being opened and closed. Everything is related to migration and environment variables so let's just consolidate that conversation here.

**Comment by thornycrackers** [comment]
I think setting the user/pass is probably not necessary . We could just have the application use the defaults of postgres/postgres and have the option of overwriting the db connection in a local_settings.py file

**Comment by alysivji** [comment]
I just hardcoded it in the docker-compose. Not sure why I did it this way, but it did not really work as intended.

**Comment by alysivji** [comment]
Closed by #84
